rust/src/http2/http2.rs

```diff
@@ -0,0 +1,1175 @@
+
+//TODOnext connection upgrade from HTTP1 cf SMTP STARTTLS
+/// C entry point for a probing parser.
+#[no_mangle]
+pub extern "C" fn rs_http2_probing_parser_tc(
+    _flow: *const Flow, _direction: u8, input: *const u8, input_len: u32, _rdir: *mut u8,
+) -> AppProto {
+    if input != std::ptr::null_mut() {
+        let slice = build_slice!(input, input_len as usize);
+        match parser::http2_parse_frame_header(slice) {
+            Ok((_, header)) => {
+                if header.reserved != 0
+                    || header.length > HTTP2_DEFAULT_MAX_FRAME_SIZE
+                    || header.flags & 0xFE != 0
+                    || header.ftype != parser::HTTP2FrameType::SETTINGS as u8
+                {
+                    return unsafe { ALPROTO_FAILED };
+                }
+                return unsafe { ALPROTO_HTTP2 };
+            }
+            Err(nom::Err::Incomplete(_)) => {
+                return ALPROTO_UNKNOWN;
+            }
+            Err(_) => {
+                return unsafe { ALPROTO_FAILED };
+            }
+        }
+    }
+    return ALPROTO_UNKNOWN;
+}
+
+#[no_mangle]
+pub extern "C" fn rs_http2_state_new() -> *mut std::os::raw::c_void {
+    let state = HTTP2State::new();
+    let boxed = Box::new(state);
+    return unsafe { transmute(boxed) };
+}
+
+#[no_mangle]
+pub extern "C" fn rs_http2_state_free(state: *mut std::os::raw::c_void) {
+    // Just unbox...
+    let mut state: Box<HTTP2State> = unsafe { transmute(state) };
+    state.free();
+}
+
+#[no_mangle]
+pub extern "C" fn rs_http2_state_tx_free(state: *mut std::os::raw::c_void, tx_id: u64) {
+    let state = cast_pointer!(state, HTTP2State);
+    state.free_tx(tx_id);
+}
+
+#[no_mangle]
+pub extern "C" fn rs_http2_parse_ts(
+    flow: *const Flow, state: *mut std::os::raw::c_void, _pstate: *mut std::os::raw::c_void,
+    input: *const u8, input_len: u32, _data: *const std::os::raw::c_void, _flags: u8,
+) -> AppLayerResult {
+    let state = cast_pointer!(state, HTTP2State);
+    let mut buf = build_slice!(input, input_len as usize);
+
+    if state.progress < HTTP2ConnectionState::Http2StateMagicDone {
+        //skip magic
+        if buf.len() >= HTTP2_MAGIC_LEN {
+            //skip magic
+            match std::str::from_utf8(&buf[..HTTP2_MAGIC_LEN]) {
+                Ok("PRI * HTTP/2.0\r\n\r\nSM\r\n\r\n") => {
+                    buf = &buf[HTTP2_MAGIC_LEN..];
+                }
+                Ok(&_) => {
+                    state.set_event(HTTP2Event::InvalidClientMagic);
+                }
+                Err(_) => {
+                    return AppLayerResult::err();
+                }
+            }
+            state.progress = HTTP2ConnectionState::Http2StateMagicDone;
+        } else {
+            //still more buffer
+            return AppLayerResult::incomplete(0 as u32, HTTP2_MAGIC_LEN as u32);
+        }
+    }
+
+    state.files.flags_ts = unsafe { FileFlowToFlags(flow, STREAM_TOSERVER) };
+    //TODOask use FILE_USE_DETECT ?
```

Comment:
What is `FILE_USE_DETECT` for ?

Generally speaking, I am not sure I get what should be done to handle files the right way
Help welcome @victorjulien

Reply:
I should revisit `FILE_USE_DETECT`. It is used to tell the API that the inspection tracker in the `File` struct should be considered when running file garbage collection.

Reply:
ok, thanks